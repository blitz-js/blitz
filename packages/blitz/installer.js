module.exports = require("./dist/installer.cjs")
