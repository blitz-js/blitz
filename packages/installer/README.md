# `installer`

The installer package houses all of the types, classes, and utilities for building a Blitz Installer. A Blitz installer is effectively just a list of steps, represented as an array of objects that conform to one of the types in the `Executor` union type (`NewFileExecutor | AddDependencyExecutor } FileTransformExecutor`). These executors are processed by the framework, executed interactively by the user, and ultimately run to install new packages to an existing Blitz app.

You can find the implementation of all Executors in the `executors/` directory, stock transforms that we'll be supplying to authors in `transforms/`, and various utilities in `utils/`, including a `paths` utility that the user can access for common paths to modify such as `_document.tsx`.
