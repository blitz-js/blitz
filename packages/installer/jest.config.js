module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  moduleFileExtensions: ["ts", "js", "json"],
  coverageReporters: ["json", "lcov", "text", "clover"],
  // collectCoverage: !!`Boolean(process.env.CI)`,
  collectCoverageFrom: ["src/**/*.ts"],
  modulePathIgnorePatterns: ["<rootDir>/tmp", "<rootDir>/dist"],
  // TODO enable threshold
  // coverageThreshold: {
  //   global: {
  //     branches: 100,
  //     functions: 100,
  //     lines: 100,
  //     statements: 100,
  //   },
  // },

  globals: {
    "ts-jest": {
      tsConfig: "test/tsconfig.json",
      isolatedModules: true,
    },
  },
}
