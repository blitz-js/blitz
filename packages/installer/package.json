{
  "name": "@blitzjs/installer",
  "version": "0.15.1",
  "description": "Package installation for the Blitz CLI",
  "homepage": "https://github.com/blitz-js/blitz#readme",
  "license": "MIT",
  "scripts": {
    "clean": "rimraf dist",
    "predev": "yarn wait:server && yarn wait:generator",
    "wait:generator": "wait-on ../generator/dist/packages/generator/src/index.d.ts",
    "wait:server": "wait-on ../server/dist/packages/server/src/index.d.ts",
    "dev": "tsdx watch --verbose",
    "build": "tsdx build",
    "test": "tsdx test",
    "test:watch": "tsdx test --watch"
  },
  "author": {
    "name": "Brandon Bayer",
    "email": "b@bayer.ws",
    "url": "https://twitter.com/flybayer"
  },
  "main": "dist/index.js",
  "module": "dist/installer.esm.js",
  "types": "dist/packages/installer/src/index.d.ts",
  "files": [
    "dist"
  ],
  "keywords": [
    "blitz",
    "installer"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/blitz-js/blitz.git"
  },
  "dependencies": {
    "@babel/core": "7.9.0",
    "@babel/plugin-transform-typescript": "7.9.4",
    "@blitzjs/display": "0.15.1",
    "@blitzjs/generator": "0.15.1",
    "ast-types": "0.13.3",
    "chokidar": "3.3.1",
    "cross-spawn": "7.0.2",
    "diff": "4.0.2",
    "enquirer": "2.3.5",
    "fs-extra": "9.0.0",
    "globby": "11.0.0",
    "ink": "2.7.1",
    "ink-link": "1.1.0",
    "ink-spinner": "3.0.1",
    "react": "experimental",
    "recast": "0.19.1",
    "ts-node": "8.9.1"
  },
  "gitHead": "d3b9fce0bdd251c2b1890793b0aa1cd77c1c0922"
}
