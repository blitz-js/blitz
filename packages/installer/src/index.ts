export * from "./recipe-executor"
export * from "./recipe-builder"
export * from "./executors/executor"
export {type as AddDependencyType} from "./executors/add-dependency-executor"
export {type as FileTransformType} from "./executors/file-transform-executor"
export {type as NewFileType} from "./executors/new-file-executor"
export {type as PrintMessageType} from "./executors/print-message-executor"

export * from "./utils/paths"
export * from "./transforms"
export {customTsParser} from "./utils/transform"
