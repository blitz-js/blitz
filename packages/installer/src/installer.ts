import {
  AddDependencyExecutor,
  isAddDependencyExecutor,
  addDependencyExecutor,
} from './executors/add-dependency-executor'
import {NewFileExecutor, isNewFileExecutor, newFileExecutor} from './executors/new-file-executor'
import {
  FileTransformExecutor,
  isFileTransformExecutor,
  fileTransformExecutor,
} from './executors/file-transform-executor'
import {log} from '@blitzjs/display'
import {logExecutorFrontmatter} from './executors/executor'
import {waitForConfirmation} from './utils/wait-for-confirmation'

type Executor = FileTransformExecutor | AddDependencyExecutor | NewFileExecutor

interface InstallerOptions {
  packageName: string
  packageDescription: string
  packageOwner: string
  packageRepoLink: string
  validateArgs?(args: {}): Promise<void>
  preInstall?(): Promise<void>
  beforeEach?(stepId: string | number): Promise<void>
  afterEach?(stepId: string | number): Promise<void>
  postInstall?(): Promise<void>
}

export class Installer<Options extends InstallerOptions> {
  private readonly steps: Executor[]
  private readonly options: Options

  constructor(options: Options, steps: Executor[]) {
    this.options = options
    this.steps = steps
  }

  private async validateArgs(cliArgs: {}): Promise<void> {
    if (this.options.validateArgs) return this.options.validateArgs(cliArgs)
  }
  private async preInstall(): Promise<void> {
    if (this.options.preInstall) return this.options.preInstall()
  }
  private async beforeEach(stepId: string | number): Promise<void> {
    if (this.options.beforeEach) return this.options.beforeEach(stepId)
  }
  private async afterEach(stepId: string | number): Promise<void> {
    if (this.options.afterEach) return this.options.afterEach(stepId)
  }
  private async postInstall(): Promise<void> {
    if (this.options.postInstall) return this.options.postInstall()
  }

  async displayFrontmatter() {
    log.branded(`Welcome to the installer for ${this.options.packageName}`)
    log.branded(this.options.packageDescription)
    log.info(`This package is authored and supported by ${this.options.packageOwner}`)
    log.info(`For additional documentation and support please visit ${this.options.packageRepoLink}`)
    console.log()
    await waitForConfirmation('Press enter to begin installation')
  }

  async run(cliArgs: {}): Promise<void> {
    await this.displayFrontmatter()
    try {
      await this.validateArgs(cliArgs)
    } catch (err) {
      log.error(err)
      return
    }
    await this.preInstall()
    for (const step of this.steps) {
      console.log() // newline

      await this.beforeEach(step.stepId)

      logExecutorFrontmatter(step)

      // using if instead of a switch allows us to strongly type the executors
      if (isFileTransformExecutor(step)) {
        await fileTransformExecutor(step, cliArgs)
      } else if (isAddDependencyExecutor(step)) {
        await addDependencyExecutor(step, cliArgs)
      } else if (isNewFileExecutor(step)) {
        await newFileExecutor(step, cliArgs)
      }

      await this.afterEach(step.stepId)
    }
    await this.postInstall()

    console.log()
    log.success(`Installer complete, ${this.options.packageName} is now be configured for your app!`)
  }
}
