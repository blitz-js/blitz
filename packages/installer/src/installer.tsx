import * as AddDependencyExecutor from "./executors/add-dependency-executor"
import * as NewFileExecutor from "./executors/new-file-executor"
import * as FileTransformExecutor from "./executors/file-transform-executor"
import {log} from "@blitzjs/display"
import {REGISTER_INSTANCE} from "ts-node"
import * as React from "react"
import {render} from "ink"
import {RecipeRenderer} from "./recipe-renderer"

export const setupTsNode = () => {
  if (!process[REGISTER_INSTANCE]) {
    // During blitz interal dev, oclif automaticaly sets up ts-node so we have to check
    require("ts-node").register({compilerOptions: {module: "commonjs"}})
  }
  require("tsconfig-paths/register")
}

type ExecutorConfig =
  | AddDependencyExecutor.Config
  | FileTransformExecutor.Config
  | NewFileExecutor.Config

export interface RecipeMeta {
  packageName: string
  packageDescription: string
  packageOwner: string
  packageRepoLink: string
  validateArgs?(args: {}): Promise<void>
  preInstall?(): Promise<void>
  postInstall?(): Promise<void>
}

export class Installer<Options extends RecipeMeta> {
  private readonly steps: ExecutorConfig[]
  private readonly options: Options

  constructor(options: Options, steps: ExecutorConfig[]) {
    this.options = options
    this.steps = steps
    setupTsNode()
  }

  // eslint-disable-next-line require-await
  private async validateArgs(cliArgs: {}): Promise<void> {
    if (this.options.validateArgs) return this.options.validateArgs(cliArgs)
  }
  // eslint-disable-next-line require-await
  private async preInstall(): Promise<void> {
    if (this.options.preInstall) return this.options.preInstall()
  }
  // eslint-disable-next-line require-await
  private async postInstall(): Promise<void> {
    if (this.options.postInstall) return this.options.postInstall()
  }

  async run(cliArgs: {}): Promise<void> {
    try {
      await this.validateArgs(cliArgs)
    } catch (err) {
      log.error(err)
      return
    }
    await this.preInstall()
    try {
      const {waitUntilExit} = render(
        <RecipeRenderer cliArgs={cliArgs} steps={this.steps} recipeMeta={this.options} />,
        {
          experimental: true,
        },
      )
      await waitUntilExit()
    } catch (e) {
      console.error(e)
    }
    await this.postInstall()

    log.info(
      `\nðŸŽ‰ The recipe for ${this.options.packageName} completed successfully! Its functionality is now fully configured in your Blitz app.\n`,
    )
  }
}
