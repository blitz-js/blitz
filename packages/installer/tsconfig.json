{
  "extends": "../../tsconfig.json",
  "include": ["src", "types", "test"],
  "exclude": ["node_modules"],
  "compilerOptions": {
    "baseUrl": "./",
    "declarationDir": "./dist",
    "downlevelIteration": true,
    "paths": {
      "*": ["src/*", "node_modules/*"]
    }
  }
}
