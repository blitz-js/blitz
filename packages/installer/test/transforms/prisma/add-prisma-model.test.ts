import {addPrismaModel} from "@blitzjs/installer"

describe("addPrismaModel", () => {
  const subject = (source: string) =>
    addPrismaModel(source, {
      type: "model",
      name: "Project",
      properties: [{type: "field", name: "id", fieldType: "String"}],
    })

  it("creates model", async () => {
    const source = `
datasource db {
  provider = "sqlite"
  url      = "file:./db.sqlite"
}`.trim()

    expect(await subject(source)).toMatchSnapshot()
  })
})
