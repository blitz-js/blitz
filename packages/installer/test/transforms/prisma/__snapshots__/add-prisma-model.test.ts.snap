// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addPrismaModel creates model 1`] = `
"
datasource db {
  provider = \\"sqlite\\"
  url      = \\"file:./db.sqlite\\"
}

model Project {
  id String
}
"
`;
